File: scripts/generate_clash.py
#!/usr/bin/python3

import os
import yaml

script_path = os.path.dirname(os.path.realpath(__file__))
os.chdir(os.path.join(script_path, '..'))

def format_ip_rules(ip_list_name):
  f = open(f'utils/ip_lists/{ip_list_name}.txt')
  rules = []
  while True:
    line = f.readline()
    if len(line) == 0:
      break
    rules.append(f'IP-CIDR,{line.strip()},DIRECT')
  f.close()
  return rules

ip_rules = format_ip_rules('lan_ip_list')
ip_rules.extend(format_ip_rules('china_ip_list'))
ip_rules.append("MATCH,SS")

clash_config = {
  "port": 7890,
  "socks-port": 7891,
  "redir-port": 0,
  "allow-lan": False,
  "mode": "Rule",
  "log-level": "info",
  "external-controller": "0.0.0.0:9090",
  "secret": "",
  "Proxy": [
    {
      "type": "ss",
      "server": "__SERVER__",
      "port": "__SERVER_PORT__",
      "password": "__PASSWORD__",
      "cipher": "__ENCRYPTION__",
      "name": "SS"
    }
  ],
  "Proxy Group": [
    {
      "name": "Proxy",
      "type": "select",
      "proxies": ["SS"]
    }
  ],
  "Rule": ip_rules
}

filepath = 'releases/clash.yml'
f = open(filepath, 'w')
f.write(yaml.dump(clash_config))
f.close()

print(f'\'{filepath}\' saved.')

File: scripts/generate_clash.rb
# clash

require 'etc'
require 'yaml'
require 'securerandom'

if Etc.getpwuid.uid == 0
  puts "Please use non-root user to execute."
  exit 1
end

PROJECT_ROOT = File.realpath(File.join(__dir__, '..'))
ASSETS_PATH = File.join(PROJECT_ROOT, 'assets')
RELEASE_PATH = File.join(PROJECT_ROOT, 'dist')
if !Dir.exist?(RELEASE_PATH)
  Dir.mkdir(RELEASE_PATH)
end

clash = {
  "port" => 7890,
  "socks-port" => 7891,
  "redir-port" => 0,
  "allow-lan" => false,
  "mode" => "Rule",
  "log-level" => "info",
  "external-controller" => "127.0.0.1:9090",
  "secret" => "",
  "Proxy" => [],
  "Proxy Group" => [
    {
      "name" => "ProxyGroup",
      "type" => "select",
      "proxies" => [] 
    }
  ],
  "Rule" => []
}

ip_list = File.read("#{ASSETS_PATH}/lan_ip_list.txt")
ip_list += File.read("#{ASSETS_PATH}/china_ip_list.txt")
ip_list.each_line do |line|
  clash["Rule"].push("IP-CIDR,#{line.strip},DIRECT")
end
clash["Rule"].push("MATCH,ProxyGroup")

filepath = "#{PROJECT_ROOT}/servers.yml"
if !File.exist?(filepath)
  File.write(filepath, File.read("#{ASSETS_PATH}/server_template.yml"))
end
proxies = []
proxy_group = []
servers = YAML.load(File.read(filepath))
servers.each do |server|
  proxy = {
    "type" => "http",
    "tls" => true,
    "name" => server["alias"],
    "server" => server["server"],
    "port" => server["port"],
    "username" => server["username"],
    "password" => server["password"],
  }
  proxies.push(proxy)
  proxy_group.push(proxy["name"])
end

clash["Proxy"] = proxies
clash["Proxy Group"][0]["proxies"] = proxy_group
filepath = "#{RELEASE_PATH}/clash.yml"
File.write(filepath, clash.to_yaml.gsub("---\n", ""))
puts "#{filepath} saved."
